### PR TITLE
Fix #10, #11 and #14

### LINKED ISSUE #10: integration should make it more clear what is going to be returned

### PR DESCRIPTION
**Pull request by zalary**
integration only returns explicitly requested fields - all the integration 'query_' fields are required (can't be deleted and must be designated before the integration will save).
integration properly counts billable items
integrations allows for [eam/email](http://intelligence.towerdata.com/eam#fields) data to be requested 

this integration will correctly handle the return of 'interests' fields but does not yet handle requesting those fields (as they are rarely returned and may presents the user with too many choices).

Deployment of this integration should be considered a breaking change, as the existing flows using this will need to be edited to specifically map the query values they want returned.

### CODE REVIEW

File: spec/outbound-spec.coffee [hunk 1]
```diff
@@ -86,6 +90,16 @@ describe 'Outbound', ->
       assert.equal outbound.request(@vars).url, 'https://api.towerdata.com/v5/td?api_key=12345&email=caitlin%40rapleafdemo.com&fields=age%2Cgender%2Ceducation%2Chousehold_income%2Chome_market_value%2Chome_owner_status%2Clength_of_residence%2Cmarital_status%2Coccupation%2Cchildren'
 
 
+  it 'should not pass towerdata not defined in the request params', ->
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: I think I parsed this name… but I'm still not sure I get what's being tested exactly. Is it that the params shouldn't have any error text in them, such as `towerdata.query_something_else` not being a real thing? Also, having two valid `query_` vars makes me wonder if that's part of the test?

**Comment by cgrayson** [comment]
Are there issues for some or all of these changes?

**Comment by zalary** [comment]
#10, #11 and #14

putting the issues in the title is problematic because then it doesn't link them.  still working on a best practices for myself for referencing those.

**Comment by cgrayson** [comment]
DUH, right there in the title, my bad. :pensive: I guess I was looking for them in that first comment.

**Comment by zalary** [comment]
no, that's what i mean - it's too hard too see.  I had to take a second look because i thought I linked them somewhere but was like, where are they?!  :stuck_out_tongue_closed_eyes:

File: src/outbound.coffee [hunk 2]
```diff
@@ -157,7 +177,10 @@ validate = (vars) ->
 response = (vars, req, res) ->
   parsedData = switch res.status
     when 200
-      event = JSON.parse(res.body)
+      orig_event = JSON.parse(res.body)
+      orig_event.email = orig_event.eam if orig_event.eam
+      delete orig_event.eam
```
Comment (cgrayson) [inline src/outbound.coffee]: I guess `delete` of a non-present thing doesn't hurt anything, but the assignment above, plus this delete could both go in an `if orig_even.eam` block, maybe.

File: src/outbound.coffee [hunk 3]
```diff
@@ -165,15 +188,19 @@ response = (vars, req, res) ->
       else
         event.outcome = 'success'
       event.annual_household_income = parseKto000(event.household_income) if event.household_income
-      event.home_market_value = parseKto000(event.home_market_value) if event.home_market_value
-      event.marital_status = parseMarital(event.marital_status) if event.marital_status
-      event.years_at_residence = parseYearsToRange(event.length_of_residence) if event.length_of_residence
-      event.residence_type = event.home_owner_status.toLowerCase() if event.home_owner_status
-      event.postal_code = event.zip if event.zip
       delete event.household_income
+      event.years_at_residence = parseYearsToRange(event.length_of_residence) if event.length_of_residence
```
Comment (cgrayson) [inline src/outbound.coffee]: Oh, these are all like that. :smiley: Less sure about that `if` block suggestion now.

File: src/outbound.coffee [hunk 4]
```diff
@@ -165,15 +188,19 @@ response = (vars, req, res) ->
...
       delete event.length_of_residence
+      event.residence_type = event.home_owner_status.toLowerCase() if event.home_owner_status
       delete event.home_owner_status
+      event.children = event.presence_of_children if event.presence_of_children
+      delete event.presence_of_children
+      event.postal_code = event.zip if event.zip
       delete event.zip
+      event.email = event.email if event.email
```
Comment (cgrayson) [inline src/outbound.coffee]: :dizzy_face:

File: src/outbound.coffee [hunk 5]
```diff
@@ -138,9 +155,12 @@ request.variables = ->
 validate = (vars) ->
   return 'email must not be blank' unless vars.lead.email
   return 'email must be valid' if vars.lead.email?.valid? and vars.lead.email.valid != true
-  return 'at least one personalization field must be designated' unless vars.towerdata?
   # prompts user to contact support since this cannot be fixed via the ui
   return 'missing API key - please contact support@activeprospect.com' unless process.env.TOWER_DATA_API_KEY?
+  ok = false
+  for key, value of vars.towerdata
+    ok = true if key.match(/^query_/) and value
+  return "at least one personalization field must be designated" unless ok
```
Comment (cgrayson) [inline src/outbound.coffee]: Boy, `Array.some` would have been nifty here, but I guess `vars.towerdata` is an object… oh well. But you might make it `break` when it finds a match.

File: spec/outbound-spec.coffee [hunk 6]
```diff
@@ -121,31 +135,45 @@ describe 'Outbound', ->
...
+            },
+          "home_owner_status": "Own",
+          "interests": {
+            "charitable_donors": true,
+            "sports": true
+            },
+          "age": "25-34",
+          "home_market_value": "150k-200k",
+          "gender": "Male",
+          "presence_of_children": "No",
```
Comment (cgrayson) [inline spec/outbound-spec.coffee]: `presence_of_children` vs `children`, that came from testing, or…?

File: src/outbound.coffee (hunk 4 above)
Comment (zalary) [inline src/outbound.coffee]: this is admittedly a little weird, but possibly related to the coffeescript null but not undefined thing we've seen before, if you don't call 'if event.email' you end up with failing tests like:

```"email": "[undefined]"```

Definitely let me know if you have any ideas on better ways to overcome that :)

File: src/outbound.coffee [hunk 7]
```diff
@@ -138,9 +155,13 @@ request.variables = ->
 validate = (vars) ->
   return 'email must not be blank' unless vars.lead.email
   return 'email must be valid' if vars.lead.email?.valid? and vars.lead.email.valid != true
-  return 'at least one personalization field must be designated' unless vars.towerdata?
   # prompts user to contact support since this cannot be fixed via the ui
   return 'missing API key - please contact support@activeprospect.com' unless process.env.TOWER_DATA_API_KEY?
+  ok = false
+  for key, value of vars.towerdata
+    ok = true if key.match(/^query_/) and value
+    break if (ok == true)
```
Comment (cgrayson) [inline src/outbound.coffee]: Just `if ok` would do it :smiley:

**Comment by cgrayson** [comment]
LGTM

**Pull request merged**